### ISSUE TITLE
Descriptive Statistics failing with pandas dataframes

### ISSUE THREAD

**Issue by KenMbuthia** [statsmodels/statsmodels#2630]
## version

statsmodels.version.full_version = '0.6.1'
python version = '3.4.3'
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.3 LTS
Release:        14.04
Codename:       trusty
## reproducing the problem

```
import numpy as np
import pandas as pd
from statsmodels.stats.descriptivestats import Describe

Describe(pd.DataFrame(np.random.randn(20).reshape((10,2)), columns=list('ab'))).summary()
```

The module statsmodels.stats.descriptivestats has the method _array_typer below.

```
def _array_typer(self):
        """if not a sctructured array"""
        if not(self.dataset.dtype.names):
            """homogeneous dtype array"""
            self._arraytype = 'homog'
        elif self.dataset.dtype.names:
            """structured or rec array"""
            self._arraytype = 'sctruct'
        else:
            assert self._arraytype == 'sctruct' or self._arraytype == 'homog'
```

The self.dataset.dtype.names on inspection of the dataset does not have dtype but has dtypes for pandas dataframe passed in as the dataset.

I'm not very sure what that method is supposed to do but it is causing AttributeError: 'DataFrame' object has no attribute 'dtype'.  Seems this summary can be a bit more descriptive than the pandas
describe which would add value. 

Hopefully this is descriptive enough and is in the right place. The problem is also in python 2.7. This is my first humble contribution to this project and in fact any open source project. Still relatively new to all of this.

**Comment by josef-pkt** [comment]
@KenMbuthia  Thanks, bug report looks good, welcome to open source

The main question is whether we should fix it or delete it. `descriptivestats` was written when pandas wasn't popular yet, and didn't have most of it's current functionality, and structured arrays were in more common use. Since pandas is now used for most datahandling, I'm not sure whether descriptivestats is still useful or necessary. It has been neglected in the last few years.

If you and others think this is still useful, then we can clean this up and fix it.

Related: we could also use it internally to provide descriptive statistics for models (endog, exog, extra data) as a convenient summary of the data.

@KenMbuthia  do you want to start to work on adding pandas support to descriptive stats, and check for other possible problems?

**Comment by KenMbuthia** [comment]
Thank you for the quick reply. 

I would like to start work on adding pandas support to descriptive stats and also check for other possible problems. To be honest, I do not have a strong background in computer science or in econometrics, my undergraduate was in bachelors of commerce. The code might not be the most efficient, fastest or most optimized and in addition I'm in the process of self-learning econometrics so some if not most all of the code I might not know what it is doing. However, I am still very interested in making my little contributions to the project. So the answer is yes. If there is something I can do to work on the project I am willing to work on it.

**Comment by bashtage** [comment]
@josef-pkt This should be dropped in favor of pandas `describe`.  Not worth fixing.

**Comment by josef-pkt** [comment]
I'm coming back to this in the context of robust estimation and heavy tailed distributions, and related to sanity checks #1908.

for example we could add standard and outlier robust methods in parallel, e.g. 
mean - median (which is also in pandas, which also includes percentiles
variance - mad
skew - robust skew
kurtosis - robust kurtosis
normality test ?
...
the moment measures could also be done under symmetric trimming, e.g. drop 10% smallest and 10% largest values
? option for transformation to normality, e.g. try box-cox if positive valued
? covariance versus gaussian rank covariance (or correlation)
? other multivariate measures (collinearity and multivariate outliers might be more in sanity checks for exog or multivariate endog)

some of the above will be more useful for endog than for the exog.

The idea for robust is to get some preliminary checks on the data that could indicate problems or issues that need investigation. A bit like a pre-estimation analog to `influence_outlier` measures.

**Comment by bksahu** [comment]
@josef-pkt  I would like to work on this. Please let me know where should I be looking into.

**Comment by josef-pkt** [comment]
I had a brief look at it last night and the code is even "older" than I thought. Most of it is oriented towards numpy structure dtypes and recarrays which have essentially been replaced in general usage by pandas DataFrames. So the code will need more refactoring than I thought initially.
I think this means it requires rewriting almost the entire class because we want to move away from recarrays and use pandas instead.

What I think would be useful

- class takes ndarray or pandas DataFrame
- split data by dtype, i.e. numeric versus string/object/non-numeric, focus on numeric columns first
- add methods/attributes for numeric, e.g. mean, var, std, percentiles, ...
  This should be extendable, so we can add robust statistics robust mean, mad and similar for scale, robust skew and kurtosis, ...
- add a `summary_frame` method that collects all statistics in a pandas DataFrame
- the above can be used for unit tests
- `summary` method should just return a SimpleTable at the end

`summary` should have list of attributes and an option to select those, e.g. "basic", "robust", and whatever else we might add.

future enhancements:
- statistics for categorical variables
- multivariate statistics, e.g. corr and robust corr
- maybe some extra methods, e.g. for identifying outlier candidates

maybe a related class with some overlap is then `stats.weightstats.DescrStatsW` which is also pretty old but not as old in style as `Describe`. `DescrStatsW` is more focused on hypothesis testing and does not include outlier robust statistics, so the duplication in functionality and use cases might not be too large to prevent having both.

**Comment by josef-pkt** [comment]
I will look a bit more at this later today.
My guess is that it will be better to start with a new class instead of fundamentally rewriting the existing Describe class.

**Comment by bksahu** [comment]
> I had a brief look at it last night and the code is even "older" than I thought. Most of it is oriented towards numpy structure dtypes and recarrays which have essentially been replaced in general usage by pandas DataFrames. So the code will need more refactoring than I thought initially.
> I think this means it requires rewriting almost the entire class because we want to move away from recarrays and use pandas instead.
> 
> What I think would be useful
> 
> * class takes ndarray or pandas DataFrame
> * split data by dtype, i.e. numeric versus string/object/non-numeric, focus on numeric columns first
> * add methods/attributes for numeric, e.g. mean, var, std, percentiles, ...
>   This should be extendable, so we can add robust statistics robust mean, mad and similar for scale, robust skew and kurtosis, ...
> * add a `summary_frame` method that collects all statistics in a pandas DataFrame
> * the above can be used for unit tests
> * `summary` method should just return a SimpleTable at the end
> 
> `summary` should have list of attributes and an option to select those, e.g. "basic", "robust", and whatever else we might add.
> 
> future enhancements:
> 
> * statistics for categorical variables
> * multivariate statistics, e.g. corr and robust corr
> * maybe some extra methods, e.g. for identifying outlier candidates
> 
> maybe a related class with some overlap is then `stats.weightstats.DescrStatsW` which is also pretty old but not as old in style as `Describe`. `DescrStatsW` is more focused on hypothesis testing and does not include outlier robust statistics, so the duplication in functionality and use cases might not be too large to prevent having both.

Okay, I would start with a new describe class in the same module and then work my way up from there.

**Comment by josef-pkt** [comment]
(another issue just reminded me)

The `__init__` for the new class should be able to use the automatic data handling to handle numpy arrays and pandas DataFrames. This would avoid extra code to do this in the class

e.g. base.model.Model has

```
    def __init__(self, endog, exog=None, **kwargs):
        missing = kwargs.pop('missing', 'none')
        hasconst = kwargs.pop('hasconst', None)
        self.data = self._handle_data(endog, exog, missing, hasconst,
                                      **kwargs)
        self.k_constant = self.data.k_constant
        self.exog = self.data.exog
        self.endog = self.data.endog
```

i.e. something like the following might work (hasconst is irrelevant)
```
self.data = self._handle_data(dataset, None, missing, hasconst,  **kwargs)
self.dataset = self.data.endog
```

this would have ynames somewhere in self.data

If this doesn't work, then I help investigating. (i.e. endog is usually univariate and I don't know if the datahandling already has full support for multivariate/multicolumn endog, but it should have most of it.)

**Comment by josef-pkt** [comment]
As a followup to this: It would be useful to get elementwise missing value handling, instead of rowwise as in the models #5416.
But I don't know yet how easy this will be beyond simple statistics like mean, std and variance.
Not needed in the first rewrite of Describe.

**Comment by bksahu** [comment]
@josef-pkt Do we need to define specific methods just to finds the stats like it is in stats.weightstats.DescrStatsW. Eg: `t0.mean` returns an array of means of all the numeric columns.

**Comment by josef-pkt** [comment]
@bksahu Yes.
I think the design is better with explicit methods. It wouldn't be necessary for the basic statistics like mean and standard deviation, but it makes it easier to extend to additional methods that are not just a single calls to a function or one-liners.

I'm not sure yet how much caching we should do, but I think we should not cache many of the method results because the usecase will be most often to just get the statistics once and use the summary_frame. 
(If we later add statistics that are more expensive to compute, then we might want to cache those.)

One question is whether the methods should return an array or a pandas Series. I guess returning a pandas Series will be more appropriate, i.e. it might be easier to combine into the summary_frame and keep track of which columns are numeric.
If the user provides data as a numpy array, then the variable index/column names could be just the column index in the numpy array.

**Comment by bksahu** [comment]
@josef-pkt I am not sure if that is what you are looking for. Could you please review it and give me feedback. Thank you.

**Comment by josef-pkt** [comment]
(parking this here so I can cleanup my checkout)

One example in the discriptivestats module fails on python 3 because of changes between string/bytes and unicode. It worked for me after switching from `S` to `U`

```
@@ -332,11 +332,11 @@ if __name__ == "__main__":
 
     data1 = np.array([(1,2,'a','aa'),
                       (2,3,'b','bb'),
                       (2,4,'b','cc')],
                      dtype = [('alpha',float), ('beta', int),
-                              ('gamma', '|S1'), ('delta', '|S2')])
+                              ('gamma', '|U1'), ('delta', '|U2')])
```

**Comment by josef-pkt** [comment]
I wanted to quickly check additional properties of the data starting with pandas describe

It's easy to just add additional things to the dataframe
(I'm not sure in the version below, do mad and iqr correspond to std or var)

```
from statsmodels.robust.scale import mad
mpgd = mpg.describe()
cv = mpg.var(0) / mpg.mean(0)
mpgd.loc["coef_var"] = cv


mpgd.loc["mad"] = mad(mpg)
mpgd.loc["std/mad"] =  mpgd.loc["std"] / mpgd.loc["mad"]
iqr = (mpgd.loc["75%"] - mpgd.loc["25%"]) / 1.349
mpgd.loc["iqr"] = iqr
mpgd.loc["std/iqr"] =  mpgd.loc["std"] / mpgd.loc["iqr"]
mpgd
````
```

print(mpgd)
          const      height       weight          hp      eng_cc
count     203.0  203.000000   203.000000  203.000000  203.000000
mean        1.0   53.733498  2555.921182  104.256158  126.857143
std         0.0    2.442864   523.205555   39.714369   41.845239
min         1.0   47.800000  1488.000000   48.000000   61.000000
25%         1.0   52.000000  2145.000000   70.000000   97.000000
50%         1.0   54.100000  2414.000000   95.000000  119.000000
75%         1.0   55.500000  2943.500000  116.000000  143.000000
max         1.0   59.800000  4066.000000  288.000000  326.000000
coef_var    0.0    0.111059   107.101915   15.128421   13.803117
mad         0.0    2.372164   578.214865   37.065055   32.617249
std/mad     NaN    1.029804     0.904864    1.071477    1.282917
iqr         0.0    2.594514   591.919941   34.099333   34.099333
std/iqr     NaN    0.941550     0.883913    1.164667    1.227157
```

- std is close to robust estimates with mad and iqr, so likely no outliers
- coefficient of variation for height is small, high is highly collinear with mean in uncentered multicollinearity measure. 
(dataset autos from GAM test case, used as example for computing condition indices and variance decomposition for #2380, and usage for #6381)

**Comment by bashtage** [comment]
Closed by afd14542d9eb32e5e355245007e9d60cb4f1b63e
